sql,colexec: fix distinct aggregation

yuzefovich opened the pull request:
**colexec: fix tuple index mismatch in ordered distinct aggregation**

We're handling the DISTINCT aggregations by populating new selection
vector to include only the tuples that are distinct to be aggregated by
a particular function. Previously, we would convert the arguments to
default aggregate function in case of ordered aggregation "with
deselection" step (this is a minor optimization). However, combining
these two things makes it very hard to know at which position in the
converted []tree.Datum vectors the arguments of a tuple actually live
which would result in us using the wrong ones. As a result, when
actually performing the aggregation we could see that there is
a selection vector (populated by the distinctness helper), yet there was
no original selection vector on the batch itself, so we would use
incorrect indices to look up the arguments in the converted []tree.Datum
slices.

Now this is fixed by following the pattern of the hash aggregator -
converting the input argument vectors without the deselection. In order
for the bug to be triggered we need to have at least two aggregate
functions with one of them handled by the default aggregate function (i.e.
it doesn't have an optimized implementation) with DISTINCT clause (if
the projection contains only that function, then the distinct operation
is pushed out of the aggregation into a distinct stage).

Fixes: #55776.

Release note (bug fix): CockroachDB could previously incorrectly compute
some aggregate functions with DISTINCT clauses when the query projects
other columns/functions and the vectorized engine was used. This bug was
introduced in 20.2.0.alpha.3 release.

**sql: fix planning of distinct stage with mix of (non-)distinct aggregations**

When performing the physical planning of aggregation we have some
optimizations:
- we try to distribute the actual aggregation if all functions support
it, the previous stage is distributed, and no function has DISTINCT
clause
- if we can't do a multi-stage aggregation, then we might plan
a distributed distinct stage if all functions have a DISTINCT clause.

However, previously, we could incorrectly evaluate whether all functions
have a DISTINCT clause if the function that doesn't support distributed
computation is preceding non-distinct function because we broke out of
the loop. As an example, queries like `SELECT corr(DISTINCT y, x),
count(y) ...` would incorrectly think that they can do local distinct
operation, but this would lead to incorrect results.

This is now fixed by separating out the optimization of performing
a local distinct stage (which is always beneficial to do; when all
functions have DISTINCT clause) and then deciding whether we can do
multi-stage aggregation.

Fixes: #55837.

Release note (bug fix): CockroachDB previously could return incorrect
results in rare circumstances when computing the aggregation functions
when some of the functions have DISTINCT clause and some don't (the
latter might not see all the necessary data), and this is now fixed.
This bug has been present for several years.

yuzefovich committed:
colexec: fix tuple index mismatch in ordered distinct aggregation

We're handling the DISTINCT aggregations by populating new selection
vector to include only the tuples that are distinct to be aggregated by
a particular function. Previously, we would convert the arguments to
default aggregate function in case of ordered aggregation "with
deselection" step (this is a minor optimization). However, combining
these two things makes it very hard to know at which position in the
converted []tree.Datum vectors the arguments of a tuple actually live
which would result in us using the wrong ones. As a result, when
actually performing the aggregation we could see that there is
a selection vector (populated by the distinctness helper), yet there was
no original selection vector on the batch itself, so we would use
incorrect indices to look up the arguments in the converted []tree.Datum
slices.

Now this is fixed by following the pattern of the hash aggregator -
converting the input argument vectors without the deselection. In order
for the bug to be triggered we need to have at least two aggregate
functions with one of them handled by the default aggregate function (i.e.
it doesn't have an optimized implementation) with DISTINCT clause (if
the projection contains only that function, then the distinct operation
is pushed out of the aggregation into a distinct stage).

Release note (bug fix): CockroachDB could previously incorrectly compute
some aggregate functions with DISTINCT clauses when the query projects
other columns/functions and the vectorized engine was used. This bug was
introduced in 20.2.0.alpha.3 release.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55780)
<!-- Reviewable:end -->

asubiotto commented:
@cathymw do you want to take a stab at reviewing this? All comments are welcome (e.g. even if they're just about misspellings) and you can also ask for an explanation for how something works if you don't understand it.

cathymw reviewed (commented):
Nice work! I just have a couple minor comments, but otherwise it looks good to me! I'll let Alfonso put the final stamp of approval though.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-:-MKBOz3vAXrQDcJBHnCH:bsocd1n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, and @yuzefovich)

___
*[pkg/sql/colexec/default_agg_tmpl.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-MKBLieC1Cy6IPZEPCE8:-MKBLieC1Cy6IPZEPCE9:b-q7ab1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/71af38fc4b3d6202d32a30826dce2314ff56db00/pkg/sql/colexec/default_agg_tmpl.go#L93)):*
> ```Go
> 		{
> 			// {{if eq "_AGGKIND" "Hash"}}
> 			// We don't need to check whether sel is non-nil in case of the
> ```

I think this comment is a little confusing because it says that `sel` doesn't need to be checked for non-nil, but it's in the else block of an if statement where `sel` was checked. I think this entire else block can be moved outside of the if statement (the way it was before). Or, alternatively, the comment can be updated to remove the part about "not needing to check `sel`".
___
*[pkg/sql/colexec/default_agg_tmpl.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-MKBNlaU-iZtg7CdFnA1:-MKBNlaU-iZtg7CdFnA2:bdm14b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/71af38fc4b3d6202d32a30826dce2314ff56db00/pkg/sql/colexec/default_agg_tmpl.go#L97)):*
> ```Go
> 			// the tuples to be aggregated.
> 			// {{end}}
> 			// Both aggregators convert the batch "sparsely" - without
> ```

Nit: I think it would be beneficial to move this comment up, since it applies to the ordered aggregator as well. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for taking a look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-:-MKBQqur-thA8cn4fbwx:bubxju8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @cathymw)

___
*[pkg/sql/colexec/default_agg_tmpl.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-MKBLieC1Cy6IPZEPCE8:-MKBQJ76CQ1qF2JQwgMX:b-k3smac) ([raw file](https://github.com/cockroachdb/cockroach/blob/71af38fc4b3d6202d32a30826dce2314ff56db00/pkg/sql/colexec/default_agg_tmpl.go#L93)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

I think this comment is a little confusing because it says that `sel` doesn't need to be checked for non-nil, but it's in the else block of an if statement where `sel` was checked. I think this entire else block can be moved outside of the if statement (the way it was before). Or, alternatively, the comment can be updated to remove the part about "not needing to check `sel`".
</blockquote></details>

Actually, this comment is inside of templated `if` block and applies only to the hash aggregation whereas the check `sel == nil` is within another templated `if` block which applies only to the ordered aggregation. The easiest way to see this is to take a look at `.eg.go` files and see what the actual generated code looks like.
___
*[pkg/sql/colexec/default_agg_tmpl.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-MKBNlaU-iZtg7CdFnA1:-MKBQfk3-qZsIQ6gAff5:bp1tb9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/71af38fc4b3d6202d32a30826dce2314ff56db00/pkg/sql/colexec/default_agg_tmpl.go#L97)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

Nit: I think it would be beneficial to move this comment up, since it applies to the ordered aggregator as well. 
</blockquote></details>

This commit does apply to be aggregators, and note that the templated `if` block for the hash aggregation ends just one line above with `{{end}}` comment.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Updated the comment to be more precise of when the bug could occur.

cathymw reviewed (commented):
Oops! Didn't realize how the templated if statements worked, TIL. Never mind then. :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-:-MKBUwl73JsdSl0NbeFc:b-3aykk8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/default_agg_tmpl.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-MKBNlaU-iZtg7CdFnA1:-MKBWzZUCtYgu4vJ4NYA:b-wpssy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/71af38fc4b3d6202d32a30826dce2314ff56db00/pkg/sql/colexec/default_agg_tmpl.go#L97)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This commit does apply to be aggregators, and note that the templated `if` block for the hash aggregation ends just one line above with `{{end}}` comment.
</blockquote></details>

meant to say "This comment does apply to both aggregators ..."


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: fix planning of distinct stage with mix of (non-)distinct aggregations

When performing the physical planning of aggregation we have some
optimizations:
- we try to distribute the actual aggregation if all functions support
it, the previous stage is distributed, and no function has DISTINCT
clause
- if we can't do a multi-stage aggregation, then we might plan
a distributed distinct stage if all functions have a DISTINCT clause.

However, previously, we could incorrectly evaluate whether all functions
have a DISTINCT clause if the function that doesn't support distributed
computation is preceding non-distinct function because we broke out of
the loop. As an example, queries like `SELECT corr(DISTINCT y, x),
count(y) ...` would incorrectly think that they can do local distinct
operation, but this would lead to incorrect results.

This is now fixed by separating out the optimization of performing
a local distinct stage (which is always beneficial to do; when all
functions have DISTINCT clause) and then deciding whether we can do
multi-stage aggregation.

Release note (bug fix): CockroachDB previously could return incorrect
results in rare circumstances when computing the aggregation functions
when some of the functions have DISTINCT clause and some don't (the
latter might not see all the necessary data), and this is now fixed.
This bug has been present for several years.

yuzefovich commented:
Found another bug, now with the physical planning of distinct aggregation in some cases, present since forever.

asubiotto reviewed (changes requested):
nit: is the 20.2 mention in the release note necessary? If so, maybe specify the point release.

I think the commit titles are again a bit too general. They need to be more specific about what they do so that you don't have to read the whole commit message to get an idea for what the change does. How about: `colexec: fix incorrect index usage in ordered distinct` for the first commit and `sql: fix distribution decision of distinct aggregations` for the second.

> Reviewed 6 of 7 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-:-MKFbnXW-GV1gG6Yc-nL:b-vwvb82)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
> nit: is the 20.2 mention in the release note necessary? If so, maybe specify the point release.

Yes, I believe it is helpful to specify the release when it was introduced, updated the release note.

> They need to be more specific about what they do so that you don't have to read the whole commit message to get an idea for what the change does.

Updated the first commit's title to be a bit more specific but didn't change the second one. I think `colexec: fix incorrect index usage in ordered distinct` is not very helpful since it's not clear what "index" it is talking about, and `sql: fix distribution decision of distinct aggregations` is actually a bit misleading - the problem is not the "distribution" itself, rather in the fact that we're planning a stage of `distinct` processors when we couldn't.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I don't want to die on this hill, but I really do care about specific commit titles. Personally, it helps me connect commits to historical context I might have in my mind ("oh this is when x fixed y and we talked about z") and identify changes at a glance (e.g. from an email subject line)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55780#-:-MKG4uSBDkwSbdIKkp93:bqk1sgw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I mean I hear your point - we don't want to have very generic commit titles since they don't provide much benefit, and I agree with that. However, I also don't want to have commit titles that are misleading about what the problem was, and sometimes it's hard to concisely express both of those things in 80 characters. I find that I prefer having a more general title that is not misleading whereas you seem to prefer it the other way around. My preference is based on the fact that if I see a general title that seems related to the question I'm trying to answer, then I'll read the commit body, but if the title is misleading, I might disregard that commit at first and will go search somewhere else mistakenly. Anyway, I updated the commit titles to be more specific.

TFTR!

bors r+

asubiotto commented:
> I also don't want to have commit titles that are misleading about what the problem was

I agree with that. If my suggestions were misleading, that's my fault. We should obviously never have misleading commit messages. It has to be the right balance of specific and brief, my issue is that a lot of commit messages lately have been too general.
